pkg/controllers/metrics/pod/controller.go

```diff
@@ -0,0 +1,194 @@
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	crmetrics "sigs.k8s.io/controller-runtime/pkg/metrics"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+const (
+	podName             = "name"
+	podNameSpace        = "namespace"
+	ownerSelfLink       = "owner"
+	podHostName         = "node"
+	podProvisioner      = "provisioner"
+	podHostZone         = "zone"
+	podHostArchitecture = "arch"
+	podHostCapacityType = "capacitytype"
+	podHostInstanceType = "instancetype"
+	podPhase            = "phase"
+	podLabels           = "podLabels"
+)
+
+var (
+	podGaugeVec = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "karpenter",
+			Subsystem: "pods",
+			Name:      "podstate",
+			Help:      "Pod state.",
+		},
+		getLabelNames(),
+	)
+)
+
+// Controller for the resource
+type Controller struct {
+	KubeClient   client.Client
+	CoreV1Client corev1.CoreV1Interface
+	LabelsMap    map[types.NamespacedName]*prometheus.Labels
+}
+
+func init() {
+	crmetrics.Registry.MustRegister(podGaugeVec)
+}
+
+func getLabelNames() []string {
+	return []string{
+		podName,
+		podNameSpace,
+		ownerSelfLink,
+		podHostName,
+		podProvisioner,
+		podHostZone,
+		podHostArchitecture,
+		podHostCapacityType,
+		podHostInstanceType,
+		podPhase,
+		podLabels,
+	}
+
+}
+
+// NewController constructs a controller instance
+func NewController(kubeClient client.Client, coreV1Client corev1.CoreV1Interface) *Controller {
+	newcontroller := Controller{
+		KubeClient:   kubeClient,
+		CoreV1Client: coreV1Client,
+		LabelsMap:    make(map[types.NamespacedName]*prometheus.Labels),
+	}
+	return &newcontroller
+}
+
+// Reconcile executes a termination control loop for the resource
+func (c *Controller) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	ctx = logging.WithLogger(ctx, logging.FromContext(ctx).Named("podmetrics").With("pod", req.Name))
+
+	// Retrieve pod from reconcile request
+	pod := &v1.Pod{}
+	if err := c.KubeClient.Get(ctx, req.NamespacedName, pod); err != nil {
+		if errors.IsNotFound(err) {
+			// Remove gauge due to pod deletion
+			if labels, ok := c.LabelsMap[req.NamespacedName]; ok {
+				podGaugeVec.Delete(*labels)
+			} else {
+				logging.FromContext(ctx).Errorf("Failed to delete gauge: failed to locate labels")
+			}
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+	// Remove the previous gauge after pod labels are updated
+	if labels, ok := c.LabelsMap[req.NamespacedName]; ok {
+		podGaugeVec.Delete(*labels)
+	}
+	newlabels, err := c.generateLabels(ctx, pod)
+	if err != nil {
+		logging.FromContext(ctx).Errorf("Failed to generate new labels: %s", err.Error())
+		return reconcile.Result{}, err
+	}
+
+	gauge, err := podGaugeVec.GetMetricWith(*newlabels)
+	if err != nil {
+		logging.FromContext(ctx).Errorf("Failed to generate new gauge: %s", err.Error())
+		return reconcile.Result{}, err
+	}
```

Comment:
Consider wrapping this in another function to avoid repeating yourself. E.g.

```
if err := c.Record(pod); err != nil {
  logging.FromContext(ctx).Errorf("Failed to record metrics, %s", err.Error())
  return reconcile.Result{}, err 
}
```
You may also want to use the noop logger to avoid controller runtime logging this twice.

Reply:
good suggestion. Done.